python/ray/air/train/data_parallel_ingest.py

```diff
@@ -76,8 +84,22 @@ def get_dataset_shards(
         for key, dataset in self.preprocessed_datasets.items():
             config = self._config(key)
 
-            if config.streamable:
-                dataset = dataset.repeat()
+            if config.use_stream_api:
+                if config.stream_window_size > 0:
+                    dataset = dataset.window(
+                        bytes_per_window=config.stream_window_size
+                    ).repeat()
+                    # In windowed mode, we re-apply the preprocessor on each iteration.
+                    if self.preprocessor:
+                        prep = self.preprocessor.transform_batch
+                        dataset = dataset.map_batches(prep, batch_format="pandas")
+                else:
+                    # If the window size is infinity, the preprocessor is cached and
+                    # we don't need to re-apply it each time.
+                    dataset = dataset.repeat()
```

Comment:
Can this logic all be done in `preprocess_datasets`? A bit confusing right now to have preprocessing logic in two different methods depending on if streaming is True.

Reply:
Nack. I think this split is more clear for now given existing need to be compatible with DatasetSpec signature.

Reply:
quite a few branches here. I wonder if the users need to care about these at all?

Reply:
From user POV it should be much easier to understand: either use batch or streaming. With streaming you can also have infinite window size as an edge case.